core/protocols/aws-query-protocol/src/main/java/software/amazon/awssdk/protocols/query/internal/unmarshall/AwsQueryResponseHandler.java

```diff
@@ -84,12 +83,10 @@ private T unmarshallResponse(SdkHttpFullResponse response) throws Exception {
      */
     private AwsResponseMetadata generateResponseMetadata(SdkHttpResponse response, Map<String, String> metadata) {
         if (!metadata.containsKey(AWS_REQUEST_ID)) {
-            metadata.put(AWS_REQUEST_ID,
-                         SdkHttpUtils.firstMatchingHeaderFromCollection(response.headers(), X_AMZN_REQUEST_ID_HEADERS)
-                                     .orElse(null));
+            metadata.put(AWS_REQUEST_ID, response.firstMatchingHeader(X_AMZN_REQUEST_ID_HEADERS).orElse(null));
         }
 
-        response.headers().forEach((key, value) -> metadata.put(key, value.get(0)));
+        response.forEachHeader((key, value) -> metadata.put(key, value.get(0)));
```

Comment:
is firstMathcingHeader still iterating on all headers?

If so can't we concider doing one pass instead?